frontend/src/metabase/visualizations/echarts/cartesian/option/trend-line.ts

```diff
@@ -0,0 +1,195 @@
+import d3 from "d3";
+// eslint-disable-next-line no-restricted-imports -- deprecated usage
+import moment from "moment-timezone";
+import _ from "underscore";
+import Color from "color";
+import type { RegisteredSeriesOption, EChartsOption } from "echarts";
+
+import { getTrendDataPointsFromInsight } from "metabase/visualizations/lib/trends";
+import type {
+  ComputedVisualizationSettings,
+  RenderingContext,
+} from "metabase/visualizations/types";
+import type { RowValue, SeriesSettings } from "metabase-types/api";
+import type { Insight } from "metabase-types/api/insight";
+
+import { applySquareRootScaling, replaceValues } from "../model/dataset";
+import type { CartesianChartModel, DataKey } from "../model/types";
+
+const TREND_LINE_DATA_KEY = "trend-line";
+
+type TrendDataset = {
+  [key: DataKey]: RowValue;
+  [TREND_LINE_DATA_KEY]: number;
+}[];
+
+/**
+ * Computes the dataset for a single series, based on its `insight` object.
+ *
+ * @param {Insight} insight - Insight object for a series
+ * @param {CartesianChartModel} chartModel - Locally computed chart data
+ * @returns {TrendDataset} Resultant dataset for the series with the given `insight` object
+ */
+function getSingleSeriesTrendDataset(
+  insight: Insight,
+  chartModel: CartesianChartModel,
+): TrendDataset {
+  const xValues = chartModel.dataset.map(
+    // We know the value is a string because it has to be a timeseries aggregation
+    row => moment(row[chartModel.dimensionModel.dataKey] as string),
+  );
+  const trendDataPoints = getTrendDataPointsFromInsight(
+    insight,
+    d3.extent(xValues),
+    xValues.length,
+  );
+
+  return trendDataPoints.map(([_x, y], rowIndex) => ({
+    [chartModel.dimensionModel.dataKey]:
+      chartModel.dataset[rowIndex][chartModel.dimensionModel.dataKey],
+    [TREND_LINE_DATA_KEY]: y,
+  }));
+}
+
+/**
+ * Normalizes the trend line datasets for each series, if the visualization is a normalized
+ * stacked bar chart. Otherwise, it will just return the input datasets without any transformation.
+ *
+ * @param {TrendDataset[]} trendDatasets - Datasets for the trend lines for each series
+ * @param {CartesianChartModel} chartModel - Locally computed chart data
+ * @param {ComputedVisualizationSettings} settings - Locally computed visualization settings for the chart
+ * @returns {TrendDataset[]} Resultant trend line datasets
+ */
+function normalizeTrendDatasets(
+  trendDatasets: TrendDataset[],
+  chartModel: CartesianChartModel,
+  settings: ComputedVisualizationSettings,
+): TrendDataset[] {
+  if (settings["stackable.stack_type"] !== "normalized") {
+    return trendDatasets;
+  }
+
+  // The chart should not be stacked if the series aren't either all bars, or all areas
+  const seriesSettings: SeriesSettings[] = chartModel.seriesModels.map(
```

Comment:
@alxnddr Perhaps we should extract this logic of checking both the `stackable.stack_type` settings value and if the series are either all bar or all area into a separate function? That way we can use it in other places related to stacking, and avoid bugs caused by forgetting this check.

Reply:
Try fetching changes from the bottom of the stack, I think after [this change](https://github.com/metabase/metabase/pull/35304/files#diff-d8d71d6406844d0b99e6222f4b3ab94bd0a7ce72e601f03f8ff2252e1e6103e9L114) you can just rely on "stackable.stack_type" and not check series display types